pkg/prometheus/agent/operator.go

```diff
@@ -550,8 +551,18 @@ func (c *Operator) addHandlers() {
 }
 
 // Resolve implements the operator.Syncer interface.
-func (c *Operator) Resolve(ss *appsv1.StatefulSet) metav1.Object {
-	key, ok := c.accessor.MetaNamespaceKey(ss)
+func (c *Operator) Resolve(obj interface{}) metav1.Object {
+	var key string
+	var ok bool
+
+	if c.daemonSetFeatureGateEnabled {
+		ds := obj.(*appsv1.DaemonSet)
+		key, ok = c.accessor.MetaNamespaceKey(ds)
+	} else {
+		ss := obj.(*appsv1.StatefulSet)
+		key, ok = c.accessor.MetaNamespaceKey(ss)
+	}
```

Comment:
Resolve() should return the PrometheusAgent object backing the statefulset or daemonset but MetaNamespaceKey() works on generic Kubernetes objects. Hence the type check should be done later to decide whether we need to call statefulSetKeyToPrometheusAgentKey() or daemonSetKeyToPrometheusAgentKey() (to be implemented).

Reply:
Yeah actually now I reread the operator's code there're still some logics missing to support updating PA DaemonSet, beside this one that you pointed out. No surprise why update e2e is failing^^ Let me try to complete these logics